### PR TITLE
Add parameters to support parsing json files with open / close tags

### PR DESCRIPTION
**Pull request by xiaomi7732**
Fix for #74

### CODE REVIEW

File: Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb [hunk 1]
```diff
@@ -117,15 +140,63 @@ def process(queue)
...
+
+          blob, content = @azure_blob.get_blob(@container, blob_name, {:start_range => start_index} )
+
+          # content will be used to calculate the new offset. Create a new variable for processed content.
+          processed_content = content
+
+          is_json_codec = @codec.is_a? LogStash::Codecs::JSON
+          if (is_json_codec)
+            skip = processed_content.index '{'
+            processed_content = processed_content[skip..-1] unless skip.nil?
```
Comment (brahmnes) [inline Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb]: start_index may start somewhere in the middle of the file. What's the reason for skipping the first '{' that you see in the middle of the file?

Also, I don't know if line 163 will cause a copy of the entire array. Since the blob is potentially large, it may be good to find ways to avoid this copying.

File: Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb [hunk 2]
```diff
@@ -135,6 +206,47 @@ def process(queue)
...
+
+    hit = 0
+    count = 0
+    index = 0
+    first = content.index('{')
+    move_opening = true
+    move_closing = true
+    while(hit < batch_size)
+        inIndex = content.index('{', index) if move_opening
+        outIndex = content.index('}', index) if move_closing
```
Comment (brahmnes) [inline Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb]: This parsing is probably not reliable.
If I have a '{' or '}' in my string values, then it will not work here. For example:

{
    "myMessage" : "I have braces like this {}"
}

File: Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb [hunk 3]
```diff
@@ -117,15 +140,63 @@ def process(queue)
...
+
+          if is_json_codec && (@break_json_down_policy != 'do_not_break')
+            @logger.debug("codec is json and policy is not do_not_break")
+            
+            @break_json_batch_count = 1 if break_json_batch_count <= 0
+            tail = processed_content[-@file_tail_bytes..-1]
+            while (!processed_content.nil? && processed_content.length > @file_tail_bytes)
+              json_event, processed_content = get_jsons(processed_content, @break_json_batch_count)
+              @logger.debug("Got json: ========================")
+              @logger.debug("#{json_event[0..50]}...#{json_event[-50..-1]}")
```
Comment (brahmnes) [inline Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb]: why 0..50? I thought this is configurable.

File: Logstash/logstash-input-azureblob/README.md [hunk 4]
```diff
@@ -56,8 +56,37 @@ When set to `start_over`, it assumes none of the blob is consumed and it will re
...
+
+Keep these parameters default to use under normal situration. Tweak these parameters when dealing with large scale azure blobs and logs.
+
+__*blob_list_page_size*__
+
+Specifies the page-size for returned blob items. Too big number will hit heap overflow; Too small number will leads to too many requests. The default of `100` is good for heap size of 1G.
+
+__*break_json_down_policy*__
+
+Only works when the codec is set to `json`. Sets the policy to break the json object in the array into small events. Break json into small sections will not be as efficient as keep it as a whole, but will reduce the usage of the memory. Possible options: `do_not_break`, `with_head_tail`, `without_head_tail`.
```
Comment (brahmnes) [inline Logstash/logstash-input-azureblob/README.md]: It would be helpful to give an example of how it breaks them down. Like how it defines the boundaries between different events.

File: Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb [hunk 5]
```diff
@@ -117,15 +140,63 @@ def process(queue)
...
+              end #if
+              @codec.decode(json_event) do |event|
+                decorate(event)
+                queue << event
+              end # decode
+            end
+          else
+            @logger.debug("Non-json codec or the policy is do not break")
+            # Putting header and content and tail together before pushing into event queue
+            processed_content = "#{header}#{processed_content}" unless header.nil? || header.length == 0
```
Comment (brahmnes) [inline Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb]: Your comment says there is a tail, but the code says it doesn't.
Also, why have the header but no tail here?

Comment (brahmnes) [reply Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb]: This is again copy the entire processed_content array, which may be expensive.

File: Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb [hunk 6]
```diff
@@ -117,15 +140,63 @@ def process(queue)
...
+            @codec.decode(processed_content) do |event|
+              decorate(event)
+              queue << event
+            end # decode
+          end #if
         ensure
           # Making sure the reader is removed from the registry even when there's exception.
           new_offset = start_index
+          new_offset = 0 if start_index == @file_head_bytes && content.nil? # Reset the offset when nothing has been read.
```
Comment (brahmnes) [inline Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb]: I don't understand this. Why resetting the offset back to the beginning of the file, just because nothing new has been written by the writer?

File: Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb [hunk 7]
```diff
@@ -135,6 +206,47 @@ def process(queue)
...
+            index = [inIndex, outIndex].min
+        end #if
+        if content[index] == '{'
+            count += 1
+            move_opening = true
+            move_closing = false
+        elsif content[index] == '}'
+            count -= 1
+            move_closing = true
+            move_opening = false
```
Comment (brahmnes) [inline Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb]: I had a really hard time trying to understand the logic for this function, and I spent 15 min starring at this function.
The nonconventional search of '{' and '}', that doesn't simply go character by character incrementally, was what made it hard to follow.

I am not sure if it would be more efficient if you loop through each character in Ruby yourself. It would be really good to add comments explaining your logic here, such as how you are search the '{' and '}', how the min function on line 233 allows you to still count the braces sequentially, and count is really for balancing the open and close brace, etc. More descriptive variable names would help a lot too.

Comment (brahmnes) [reply Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb]: Having said that, I think your parsing logic can still fail if the literal string has a brace inside it. See if there are existing JSON parser that can do the job.

File: Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb (hunk 1 above)
Comment (xiaomi7732) [reply Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb]: @brahmnes,
1. You read it correct that it might start in the middle except that it skips any characters before the first `{`. 
2. I can look into the copy content issue. That is a valid consideration for hyper-scale scenario, although by benchmark, I don't see it blown up memory for logs at the size of 300MB.

File: Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb (hunk 2 above)
Comment (xiaomi7732) [reply Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb]: Very good point. I'll file a issue to track this / fix this issue.

File: Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb (hunk 3 above)
Comment (xiaomi7732) [reply Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb]: Are you asking json_even[0..50]...json_event[-50..-1]? That is just log output for debugging. I would like to see the start / end of the json event parsed, because if there's something wrong, it will like happen there near the beginning or ending.

File: Logstash/logstash-input-azureblob/README.md (hunk 4 above)
Comment (xiaomi7732) [reply Logstash/logstash-input-azureblob/README.md]: Good idea. Worth a wiki page for this.

File: Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb (hunk 5 above)
Comment (xiaomi7732) [reply Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb]: Thanks for pointing it out. The tail is, in fact, there in `processed_content`, which is why worth a comment :-)

File: Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb (hunk 7 above)
Comment (xiaomi7732) [reply Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb]: For question 1, I benchmarked the performance, the current implementation is around 30 times faster than going through the string 1 by another character in typical cases.

That said, it is a good idea to introduce native json parser, the current catch is that json parser usually asks a well formed json. I'll keep looking into it.

File: Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb (hunk 6 above)
Comment (xiaomi7732) [reply Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb]: Yes, this is an edge case. If the start index is at the header while there is no content in the body, basically mean there's something wrong - breaking in the process / exception thrown / malformed json file and so on - which should not happen in normal case. I reset the index that the file will be treated as a non-touched file that it will be picked up and try again in the next iteration.

File: Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb (hunk 1 above)
Comment (brahmnes) [reply Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb]: Got your point about skipping to the first '{'.
For #2, the memory is a concern, but the bigger concern is speed. If this does get up to several hundreds of MB, and your agent is trying to process many of these blobs, then it can make a difference.

File: Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb (hunk 3 above)
Comment (brahmnes) [reply Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb]: I misread your code earlier when I added the comment, so I take back what I said.
However, if your json_event is less than 50 characters long, would this throw an exception?

File: Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb (hunk 6 above)
Comment (brahmnes) [reply Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb]: ok, reasonable

File: Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb (hunk 3 above)
Comment (xiaomi7732) [reply Logstash/logstash-input-azureblob/lib/logstash/inputs/azureblob.rb]: Good concern on exception thrown. No, it won't thrown. It might output the same json twice.
